### ISSUE TITLE
[1.12.0rc5] files deleted in previous layers are still present in later layers with overlay2

### ISSUE THREAD

**Issue by chriskuehl** [docker/docker#25193]
**Output of `docker version`:**

```
Client:
 Version:      1.12.0-rc5
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   a3f2063
 Built:        Tue Jul 26 13:43:25 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-rc5
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   a3f2063
 Built:        Tue Jul 26 13:43:25 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 5
Server Version: 1.12.0-rc5
Storage Driver: overlay2
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host bridge null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-28-generic
Operating System: Ubuntu 16.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 992.4 MiB
Name: ubuntu-xenial
ID: 5523:OQEI:KAR4:FPLJ:ZUB4:GZI5:UJE7:JHZY:GA2M:BOKN:QEV7:OSMF
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
This happens on AWS and VirtualBox (vagrant).

**Steps to reproduce the issue:**
Here is a vagrant reproduction:
```
$ vagrant init ubuntu/xenial64
$ vagrant up
$ vagrant ssh

# install Docker 1.12rc5
$ curl -fsSL https://test.docker.com/ | sh
$ sudo systemctl stop docker
$ sudo dockerd -H fd:// -s overlay2 &
$ sudo docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 1.12.0-rc5
Storage Driver: overlay2
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host bridge null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-28-generic
Operating System: Ubuntu 16.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 992.4 MiB
Name: ubuntu-xenial
ID: 5523:OQEI:KAR4:FPLJ:ZUB4:GZI5:UJE7:JHZY:GA2M:BOKN:QEV7:OSMF
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8


# replicate bug
$ cat > Dockerfile <<EOF
FROM debian:jessie
RUN mkdir -p /test/wat
RUN touch /test/a /test/b /test/c /test/d
RUN rm -r /test/a /test/b /test/c /test/d /test/wat
RUN ls -l /test
EOF
$ sudo docker build --no-cache -t wat .
```

**Describe the results you received:**
Files deleted in a previous layer still exist.

```
Sending build context to Docker daemon 15.87 kB
Step 1 : FROM debian:jessie
jessie: Pulling from library/debian
5c90d4a2d1a8: Pull complete 
Digest: sha256:8b1fc3a7a55c42e3445155b2f8f40c55de5f8bc8012992b26b570530c4bded9e
Status: Downloaded newer image for debian:jessie
 ---> 1b088884749b
Step 2 : RUN mkdir -p /test/wat
 ---> Running in b3ff0269cb42
 ---> ccd645731e04
Removing intermediate container b3ff0269cb42
Step 3 : RUN touch /test/a /test/b /test/c /test/d
 ---> Running in 32b39936e50c
 ---> 883f4b62ddd2
Removing intermediate container 32b39936e50c
Step 4 : RUN rm -r /test/a /test/b /test/c /test/d /test/wat
 ---> Running in d99a4f85166e
 ---> 93d7f01dfef8
Removing intermediate container d99a4f85166e
Step 5 : RUN ls -l /test
 ---> Running in 64e842cd4f13
total 4
-rw-r--r-- 1 root root    0 Jul 28 19:03 a
-rw-r--r-- 1 root root    0 Jul 28 19:03 b
-rw-r--r-- 1 root root    0 Jul 28 19:03 c
-rw-r--r-- 1 root root    0 Jul 28 19:03 d
drwxr-xr-x 2 root root 4096 Jul 28 19:03 wat
```

**Describe the results you expected:**
Files deleted in a previous layer shouldn't still exist.

```
Sending build context to Docker daemon 15.87 kB
Step 1 : FROM debian:jessie
jessie: Pulling from library/debian
5c90d4a2d1a8: Pull complete 
Digest: sha256:8b1fc3a7a55c42e3445155b2f8f40c55de5f8bc8012992b26b570530c4bded9e
Status: Downloaded newer image for debian:jessie
 ---> 1b088884749b
Step 2 : RUN mkdir -p /test/wat
 ---> Running in b3ff0269cb42
 ---> ccd645731e04
Removing intermediate container b3ff0269cb42
Step 3 : RUN touch /test/a /test/b /test/c /test/d
 ---> Running in 32b39936e50c
 ---> 883f4b62ddd2
Removing intermediate container 32b39936e50c
Step 4 : RUN rm -r /test/a /test/b /test/c /test/d /test/wat
 ---> Running in d99a4f85166e
 ---> 93d7f01dfef8
Removing intermediate container d99a4f85166e
Step 5 : RUN ls -l /test
 ---> Running in 64e842cd4f13
total 0
```

This seems to work in overlay and aufs, but not overlay2.

**Comment by cpuguy83** [comment]
ping @dmcgowan

**Comment by dmcgowan** [comment]
Fixed by #25086, included in 1.12.0 final
